build: Bump minimum libc to 2.17 for release binaries

fanquake opened the pull request:
Closes: #17525. Taken over from #17531. 

Debian 8 (Jessie) has:
- g++ version 4.9.2
- libc version 2.19

CentOS 7 has:
- g++ version 4.8.5
- libc version 2.17

Ubuntu 16.04.4 (Xenial, oldest supported Ubuntu) has:
- g++ version 5.3.1
- libc version 2.23.0

Taking the minimum of these as our target. According to [GNU ABI document](https://gcc.gnu.org/onlinedocs/libstdc++/manual/abi.html) this corresponds to:

- GCC 4.8.5: GCC_4.8.0
- (glibc)    GLIBC_2_17

This also contains a (long needed) commit to disallow dynamic linking to stdc++, as our releases statically link against that.

laanwj committed:
build: Disallow dynamic linking against c++ library

Ever since statically linking Qt, we've been linking the C++ library
statically too (-static-libstdc++). Take this into
account in the symbol checker.

laanwj committed:
build: Bump minimum versions in symbol checker

Debian 8 (Jessie) has:
- g++ version 4.9.2
- libc version 2.19

Ubuntu 16.04.4 (Xenial) has:
- g++ version 5.3.1
- libc version 2.23.0

CentOS 7 has:
- g++ version 4.8.5
- libc version 2.17

Taking the minimum of these as our target.
According to the GNU ABI document this corresponds to:

- GCC 4.8.5: GCC_4.8.0
- (glibc)    GLIBC_2_17

Co-Authored-By: fanquake <fanquake@gmail.com>

MarcoFalke commented on contrib/devtools/symbol-check.py:
```diff
@@ -16,30 +16,30 @@
+# - g++ version 5.3.1 (https://packages.ubuntu.com/search?keywords=g%2B%2B&searchon=names&suite=xenial&section=all)
+# - libc version 2.23.0 (https://packages.ubuntu.com/search?keywords=libc6&searchon=names&suite=xenial&section=all)
+#
+# CentOS 7 has:
```
```suggestion
# CentOS 7. EOL: June 30, 2024. https://wiki.centos.org/FAQ/General#head-fe8a0be91ee3e7dea812e8694491e1dde5b75e6d
```

MarcoFalke commented on contrib/devtools/symbol-check.py:
```diff
@@ -16,30 +16,30 @@
 import os
 
-# Debian 6.0.9 (Squeeze) has:
+# Debian 8 (Jessie) has:
```
```suggestion
# Debian 8 Jessie. EOL: 2020. https://wiki.debian.org/DebianReleases#Production_Releases
```

MarcoFalke commented on contrib/devtools/symbol-check.py:
```diff
@@ -16,30 +16,30 @@
+# - libc version 2.19 (https://packages.debian.org/search?suite=jessie&arch=any&searchon=names&keywords=libc6)
 #
-# Ubuntu 10.04.4 (Lucid Lynx) has:
+# Ubuntu 16.04 (Xenial) has:
```
```suggestion
# Ubuntu 16.04 Xenial. EOL: April, 2021. https://help.ubuntu.com/community/EOL#Ubuntu_16.04_Xenial_Xerus
```

flack replied:
@MarcoFalke there seems to be some ambiguity about the EOL date for 16.04, this page says it's April 2024, and April 2021 is only the "End of Standard Support":

https://wiki.ubuntu.com/Releases

fanquake replied:
I've changed it to use the 2024 date.

maflcko reviewed (approved):
ACK

Could mention the EOL dates and remove all the links from the OP? They don't need to end up in the git merge log, imo.

sidhujag commented:
concept ACK can we remove some fallbacks from src/compat/glibc_compat.cpp?

MarcoFalke commented:
Failure:

```
  GEN      qt/locale/bitcoin_zh_TW.qm
  GEN      qt/qrc_bitcoin_locale.cpp
  CXX      qt/qt_libbitcoinqt_a-qrc_bitcoin_locale.o
  AR       qt/libbitcoinqt.a
  CXXLD    qt/bitcoin-qt
make[2]: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-linux-gnu/src'
make[1]: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-linux-gnu/src'
Making all in doc/man
make[1]: Entering directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-linux-gnu/doc/man'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-linux-gnu/doc/man'
make[1]: Entering directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-linux-gnu'
make[1]: Nothing to be done for 'all-am'.
make[1]: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-linux-gnu'
+ make -j1 -C src check-security
make: Entering directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-linux-gnu/src'
Checking binary security...
make: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-linux-gnu/src'
+ make -j1 -C src check-symbols
make: Entering directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-linux-gnu/src'
Checking glibc back compat...
bitcoin-cli: export of symbol environ not allowed
bitcoin-cli: export of symbol _environ not allowed
bitcoin-cli: export of symbol __environ not allowed
bitcoin-tx: export of symbol environ not allowed
bitcoin-tx: export of symbol _environ not allowed
bitcoin-tx: export of symbol __environ not allowed
bitcoin-wallet: export of symbol environ not allowed
bitcoin-wallet: export of symbol _environ not allowed
bitcoin-wallet: export of symbol __environ not allowed
test/test_bitcoin: export of symbol _environ not allowed
test/test_bitcoin: export of symbol environ not allowed
test/test_bitcoin: export of symbol __environ not allowed
qt/bitcoin-qt: export of symbol _environ not allowed
qt/bitcoin-qt: export of symbol environ not allowed
qt/bitcoin-qt: export of symbol __environ not allowed
Makefile:15062: recipe for target 'check-symbols' failed
make: *** [check-symbols] Error 1
make: Leaving directory '/home/ubuntu/build/bitcoin/distsrc-x86_64-linux-gnu/src'

fanquake commented:
I've updated this to include the dates, and removed the links from the PR / commit bodies.

> concept ACK can we remove some fallbacks from src/compat/glibc_compat.cpp?

If we decide to do that, it wont be done as part of this PR.

> bitcoin-cli: export of symbol environ not allowed
> bitcoin-cli: export of symbol _environ not allowed

Will look into this now.

MarcoFalke commented on contrib/devtools/symbol-check.py:
```diff
@@ -16,30 +16,30 @@
+# - libc version 2.19 (https://packages.debian.org/search?suite=jessie&arch=any&searchon=names&keywords=libc6)
 #
-# Ubuntu 10.04.4 (Lucid Lynx) has:
+# Ubuntu 16.04 (Xenial) EOL: 2024. https://help.ubuntu.com/community/EOL#Ubuntu_16.04_Xenial_Xerus
```
```suggestion
# Ubuntu 16.04 (Xenial) EOL: 2024. https://wiki.ubuntu.com/Releases
```

laanwj commented:
ACK

>  Failure:

This only increases the allowes versions in the symbol checker. I'm confused how it could make it detect a new exported symbol. Do you build with the same configuration as the gitian descriptors?

MarcoFalke commented:
> Do you build with the same configuration as the gitian descriptors

It should use the gitian descriptor of each commit, but they weren't changed in between, so it shouldn't matter either way.

laanwj commented:
So the changes here cause an exported (not imported) symbol to be flagged? I'm really confused.

fanquake committed:
doc: add glibc 2.17 requirement to release-notes

fanquake commented:
Discussed with @laanwj and the `environ` issue is going to be fixed separately from increasing the minimum required `glibc` version.

Rebased, added a release note and fixed the nit from above.

laanwj commented:
re-ACK 8f15a317602727d24f60d0bbf43f851b33df3228

hebasto commented:
@fanquake @laanwj 

I'm confused with the commit 16a7be1663b02ddefa1e4f0309be49b725ffb388 "build: Bump minimum versions in symbol checker".

The commit message and PR title state that minimum version of glibc is 2.17 from now.
From release notes:
> glibc 2.17 or greater is now required to run the release binaries.

But the name of [`MAX_VERSIONS` dictionary](https://github.com/bitcoin/bitcoin/blob/ffad34816722cdf27a0a7c16539ddd1d655602e0/contrib/devtools/symbol-check.py#L40-L44), and the `ver <= max_versions[lib]` in https://github.com/bitcoin/bitcoin/blob/ffad34816722cdf27a0a7c16539ddd1d655602e0/contrib/devtools/symbol-check.py#L175 convince me that 2.17 is the maximum allowed version of glibc symbols.

Please help me to get the right idea on this.
